### PR TITLE
Part of MDEV-30645: Make create trigger a synonym for create event

### PR DESCRIPTION
**Pull request by andremralves**
- [x] *The Jira issue number for this PR is: MDEV-30645*

## Description

This PR makes `CREATE TRIGGER `a synonym for `CREATE EVENT`, e.g.:

```sql
CREATE EVENT myevent
    ON SCHEDULE AT CURRENT_TIMESTAMP + INTERVAL 1 HOUR
    DO
      UPDATE myschema.mytable SET mycol = mycol + 1;
```

### Implementation:

I changed the rule `trigger_tail` to be either a `trigger_tail` or a `event_tail`. To do so, I had to change position of the checking:
```cpp
{
  if (unlikely(Lex->add_create_options_with_check($2)))
    MYSQL_YYABORT;
}
```
putting it a little bit lower to prevent the following error:
```bash
[4/36] [BISON][gen_mariadb_cc_hh] Building parser with bison 3.8.2
FAILED: sql/yy_mariadb.cc sql/yy_mariadb.hh /home/andre/repos/server/sql/yy_mariadb.cc /home/andre/repos/server/sql/yy_mariadb.hh 
cd /home/andre/repos/server/sql && /usr/bin/bison -p MYSQL -S /home/andre/repos/server/sql/myskel.m4 -d -o /home/andre/repos/server/sql/yy_mariadb.cc /home/andre/repos/server/sql/yy_mariadb.yy
/home/andre/repos/server/sql/yy_mariadb.yy: error: shift/reduce conflicts: 502 found, 62 expected
/home/andre/repos/server/sql/yy_mariadb.yy: note: rerun with option '-Wcounterexamples' to generate conflict counterexamples
/home/andre/repos/server/sql/yy_mariadb.yy:18198.11-18201.11: warning: rule useless in parser due to conflicts [-Wother]
18198 |           {
      |           ^

```

## Release Notes
TODO: What should the release notes say about this change?
Include any changed system variables, status variables or behaviour. Optionally list any https://mariadb.com/kb/ pages that need changing.

## How can this PR be tested?

You can use the automated test:
```
./mtr main.create_trigger_on_schedule 
```
or manually try to create an event using `CREATE TRIGGER` and then check if there is any corner case.

TODO: modify the automated test suite to verify that the PR causes MariaDB to behave as intended.
Consult the documentation on ["Writing good test cases"](https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server).

If the changes are not amenable to automated testing, please explain why not and carefully describe how to test manually.

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch.*
- [ ] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

### CODE REVIEW

**Comment by vuvova** [comment]
Looks good, but please add a test for a syntax error, like
```sql
CREATE OR REPLACE TRIGGER IF NOT EXISTS ev1 IN SCHEDULE EVERY 1 SECOND DO DROP DATABASE db1;
```

**Comment by andremralves** [comment]
@vuvova  I added some more tests like the one you suggested.

File: mysql-test/main/create_trigger_on_schedule.result [hunk 1]
```diff
@@ -0,0 +1,59 @@
+SET timestamp=UNIX_TIMESTAMP('2014-11-01 10:20:30');
+SET GLOBAL event_scheduler=off;
+CREATE TABLE t1 (a INT);
+CREATE OR REPLACE TRIGGER IF NOT EXISTS ev1 ON SCHEDULE EVERY 1 SECOND DO DROP DATABASE db1;
+ERROR HY000: Incorrect usage of OR REPLACE and IF NOT EXISTS
+CREATE OR REPLACE TRIGGER IF NOT EXISTS ev1 IN SCHEDULE EVERY 1 SECOND DO DROP DATABASE db1;
```
Comment (vuvova) [inline mysql-test/main/create_trigger_on_schedule.result:6]: That's not very useful, is it? The parser aborts at `IF NOT EXISTS` it doesn't even read your `IN`.

**Review by vuvova — CHANGES_REQUESTED** [review]

File: mysql-test/main/create_trigger_on_schedule.result (hunk 1 above)
Comment (andremralves) [reply mysql-test/main/create_trigger_on_schedule.result]: Hi, sorry about that. Actually, I noticed the error was a odd, but I thought maybe it was intended. I updated the command, now showing a parser error. Does it look good now?

Comment (vuvova) [reply mysql-test/main/create_trigger_on_schedule.result]: Not really. Now you have
```sql
CREATE OR REPLACE EVENT ev1 IN SCHEDULE EVERY 1 SECOND DO DROP DATABASE db1;
```
which correctly fails with a syntax error, as it always did. You implemented `CREATE TRIGGER` but you test here `CREATE EVENT`, old syntax.

Comment (andremralves) [reply mysql-test/main/create_trigger_on_schedule.result]: Ahh, true. I don't know how didn't saw that or even how `EVENT` appeared there in the first place, I'm sure I was using trigger. :laughing:

**Review by vuvova — APPROVED** [review]
Looks good, thanks.

It's a bit inconsistent that error messages still say "event", like in
```
CREATE TRIGGER ev1 ON SCHEDULE EVERY 1 SECOND DO INSERT INTO t1 VALUES (11);
ERROR HY000: Event 'ev1' already exists
```
but let's not spend time on it now. You can get back to it after the main GSoC project is done.

**Comment by vuvova** [comment]
Note, this won't be pushed into 11.6 right now, it would just be seen as a strange and redundant alias of an existing feature.

We'll push everything together, then it'll be a logical part of the "generalized triggers"

**Comment by andremralves** [comment]
> Looks good, thanks.
> 
> It's a bit inconsistent that error messages still say "event", like in
> 
> ```
> CREATE TRIGGER ev1 ON SCHEDULE EVERY 1 SECOND DO INSERT INTO t1 VALUES (11);
> ERROR HY000: Event 'ev1' already exists
> ```
> 
> but let's not spend time on it now. You can get back to it after the main GSoC project is done.

That is something I was thinking about. After we use `CREATE TRIGGER .. ON`, should the outcome be an event or a trigger? The name we use is trigger, but all the characteristics are from an event. Also, when we use `SHOW CREATE EVENTS`, should the ones created with `CREATE TRIGGER` appear as well? With this implementation, `CREATE TRIGGER .. ON` will just be another way to create an event. Would it be better to differentiate them?

**Comment by vuvova** [comment]
I think it'll be the same object, so no need to differentiate. `EVENT` is very overloaded concept, e.g. very often people speak of "binlog events", which are completely different. What you're doing creates a notion of a "generalized trigger", like in some other databases. The syntax is `CREATE TRIGGER <name> <condition> ...` and the condition can be "AFTER INSERT"  or "BEFORE SHUTDOWN" or whatever, may be "ON LOGIN" or "ON SLAVE CONNECT", etc. In this scheme "ON SCHEDULE" is just another condition for a trigger. And `CREATE EVENT` is a historical syntax to create a scheduled generalized trigger.

But it should be done in steps.

**Comment by andremralves** [comment]
Thanks for the explanation. I thought that since they are going to be the same object (event), maybe the error message saying it is an event is good, but, in the other hand, it can be confusing for the user.

**Review by vuvova — CHANGES_REQUESTED** [review]
looks very good so far

File: mysql-test/main/create_trigger_after_startup.test [hunk 2]
```diff
@@ -0,0 +1,16 @@
+--source include/not_embedded.inc
+--source include/no_view_protocol.inc
```
[the remark this answers is not in the document]
Comment (andremralves) [reply mysql-test/main/create_trigger_after_startup.test:2]: I thought starting with this was standard. I will remove then.

File: mysql-test/main/create_trigger_on_schedule.result [hunk 3]
```diff
@@ -0,0 +1,59 @@
+SET timestamp=UNIX_TIMESTAMP('2014-11-01 10:20:30');
```
[the remark this answers is not in the document]
Comment (andremralves) [reply mysql-test/main/create_trigger_on_schedule.result:1]: I made this test based on `create_drop_event.test` just to check the same operations, but using the new syntax `CREATE TRIGGER`. So I don't know why timestamp is set, but it's also set in `create_drop_event.test`. Maybe it's important for something, but I can remove for now.

File: sql/event_scheduler.cc [hunk 4]
```diff
@@ -837,6 +837,26 @@ Event_scheduler::dump_internal_status()
...
+    FALSE  OK
+    TRUE   Error (Serious error)
+*/
+
+bool
+Event_scheduler::run_single(Event_queue_element_for_exec *event_name)
+{
+  DBUG_ENTER("Event_scheduler::run_single");
+  
+  DBUG_RETURN(execute_top(event_name));
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/event_scheduler.cc:857]: It was my first implementation -- making `execute_top` public -- and though you would say the opposite :laughing:

File: mysql-test/main/create_trigger_after_startup.test (hunk 2 above)
[the remark this answers is not in the document]
Comment (vuvova) [reply mysql-test/main/create_trigger_after_startup.test:2]: many test files don't have that, so, definitely not standard.

The first line makes the test to be skipped when run like `./mtr --embedded testname`.
The second line makes the test to be skipped for `./mtr --view-protocol`

Most tests run fine both in `--embedded` and in `--view`. Every case when some test doesn't needs to be analyzed, it might be a bug or might be a fundamental limitation of a feature, then you add a line like above.

I suspect that your test will work fine in `--view` (try it). But might indeed not work in `--embedded` (test it too)

File: mysql-test/main/create_trigger_on_schedule.result (hunk 3 above)
[the remark this answers is not in the document]
Comment (vuvova) [reply mysql-test/main/create_trigger_on_schedule.result:1]: this "fixes time" inside the client connection. Needed when the following tests will show it. In `create_drop_event.test` it's not needed either.

If you'd do `SELECT * INFORMATION_SCHEMA.EVENTS` for example, it has columns for "CREATED", "LAST_ALTERED", "LAST_EXECUTED" and if you don't fix the timestamp first, you won't get a repeatable result (unless you replace these values away).

File: mysql-test/main/mysql_upgrade-28915.result [hunk 5]
```diff
@@ -38,6 +38,7 @@ event	CREATE TABLE `event` (
   `body` longblob NOT NULL,
   `definer` varchar(384) CHARACTER SET utf8mb3 COLLATE utf8mb3_bin NOT NULL DEFAULT '',
   `execute_at` datetime DEFAULT NULL,
+  `db_event` enum('AFTER_STARTUP','BEFORE_SHUTDOWN') DEFAULT NULL,
```
[the remark this answers is not in the document]
Comment (andremralves) [reply mysql-test/main/mysql_upgrade-28915.result]: Fixed!

File: sql/event_data_objects.h [hunk 6]
```diff
@@ -96,6 +96,7 @@ class Event_queue_element : public Event_basic
 public:
   int on_completion;
   int status;
+  int dbevent;
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/event_data_objects.h]: Right, I will go with `db_event`. I was using `dbname` in `Event_parse_data` as a reference, but then I saw some tables using  `db_name`. That is why I ended up using both names, `dbevent` in the code and `db_event` in the table. I think `db_event` is the right choice.

File: sql/event_data_objects.cc [hunk 7]
```diff
@@ -532,6 +540,31 @@ Event_queue_element::load_from_row(THD *thd, TABLE *table)
...
+    String str(buff, sizeof(buff), &my_charset_bin);
+    LEX_CSTRING tmp;
+
+    table->field[ET_FIELD_DB_EVENT]->val_str(&str);
+    if (!(tmp.length= str.length()))
+      DBUG_RETURN(TRUE);
+
+    tmp.str= str.c_ptr_safe();
+
+    i= find_string_in_array(enum_dbevent_to_name, &tmp, system_charset_info);
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/event_data_objects.cc]: Hmm, much simpler. I was wondering if I could do it this way without any problem.

Would it be possible to do the same for ET_FIELD_TRANSIENT_INTERVAL here?
https://github.com/MariaDB/server/blob/d83742622dd1bece3f6a3a11ac0be64de0f3ff84/sql/event_data_objects.cc#L541-L558

I thought maybe there is some corner case I'm not seeing that requires me to implement something like the above code.

File: sql/events.cc [hunk 8]
```diff
@@ -1303,6 +1311,97 @@ Events::load_events_from_db(THD *thd)
   DBUG_RETURN(ret);
 }
 
+bool
+Events::search_n_execute_db_events(THD *thd,
+                                   Event_parse_data::enum_dbevent dbevent) {
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/events.cc]: Fixed!

File: sql/events.cc [hunk 9]
```diff
@@ -1303,6 +1311,97 @@ Events::load_events_from_db(THD *thd)
...
+  DBUG_ENTER("Events::search_n_execute_db_events");
+  /* We need a temporary THD */
+  if (!thd)
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/events.cc]: I thought maybe I would need to pass it as a parameter on another occasion. Like in `Event::init`.

File: sql/events.cc [hunk 10]
```diff
@@ -1303,6 +1311,97 @@ Events::load_events_from_db(THD *thd)
...
+                   "Event Scheduler: Failed to open table mysql.event",
+                   MYF(ME_ERROR_LOG));
+    DBUG_RETURN(TRUE);
+  }
+
+  if (init_read_record(&read_record_info, thd, table, NULL, NULL, 0, 1, FALSE))
+  {
+    close_thread_tables(thd);
+    DBUG_RETURN(TRUE);
+  }
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/events.cc:1357]: I will try to use it.

[the remark this answers is not in the document]
Comment (andremralves) [reply sql/events.cc:1357]: Actually, I think it won't work. Events::load_events_from_db loads events into the event_queue, but I'm not using the event_queue for startup and shutdown triggers. I'm only loading SCHEDULE events into the event_queue to avoid any conflict in the event_scheduler code, since startup and shutdown triggers don't work based on time.

Should I make Events::load_events_from_db a two-in-one method that loads events into the event_queue or executes some generalized triggers?

[the remark this answers is not in the document]
Comment (vuvova) [reply sql/events.cc:1357]: okay, if it won't make the code simpler — don't

**Review by vuvova — CHANGES_REQUESTED** [review]

File: sql/event_data_objects.cc [hunk 11]
```diff
@@ -317,6 +318,7 @@ Event_queue_element::Event_queue_element():
 
   starts= ends= execute_at= last_executed= 0;
   starts_null= ends_null= execute_at_null= TRUE;
+  event_kind= 0;
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/event_data_objects.cc:321]: Okay, I agree. I made it an int to match on_completion and status, as they are also integers but receive values based on enums. Should I change `on_completion` and `status` as well?

File: sql/sql_yacc.yy [hunk 12]
```diff
@@ -13193,8 +13247,19 @@ drop:
...
+            /*
+                If the trigger does not exist, try to drop a
+                trigger from mysql.event
+            */
+            char trn_path_buff[FN_REFLEN];
+            LEX_CSTRING trn_path= { trn_path_buff, 0 };
+            build_trn_path(thd, lex->spname, (LEX_STRING*) &trn_path);
+            if (check_trn_exists(&trn_path) && !opt_bootstrap)
+              lex->set_command(SQLCOM_DROP_EVENT, $3);
+#endif
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/sql_yacc.yy:13262]: What do you suggest? Maybe make this decision in `mysql_create_or_drop_trigger()`?

[the remark this answers is not in the document]
Comment (andremralves) [reply sql/sql_yacc.yy:13262]: The idea is to execute `drop event` if a trigger isn't found, right?

File: sql/mysqld.cc [hunk 13]
```diff
@@ -5988,6 +5994,8 @@ int mysqld_main(int argc, char **argv)
   if (Events::init((THD*) 0, opt_noacl || opt_bootstrap))
     unireg_abort(1);
 
+  Events::search_n_execute_events(Event_parse_data::STARTUP);
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/mysqld.cc:5997]: Okay

File: sql/sql_yacc.yy (hunk 12 above)
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/sql_yacc.yy:13262]: What do you think about handling it in sql_parser.cc? First, DROP TRIGGER is executed, and then, if it returns an error, we try to execute DROP EVENT. Like this:

```cpp
  case SQLCOM_DROP_TRIGGER:
  {
    if (thd->variables.option_bits & OPTION_IF_EXISTS)
      lex->create_info.set(DDL_options_st::OPT_IF_EXISTS);

    /* Conditionally writes to binlog. */
    res= mysql_create_or_drop_trigger(thd, all_tables, 0);

    /* 
       If there is an error or error level warning
       try to drop from mysql.event.
    */
    if(!opt_bootstrap &&
       (thd->get_stmt_da()->sql_errno() + thd->get_stmt_da()->error_count()))
    {
      /* Clear errors and warnings */
      thd->clear_error();
      thd->get_stmt_da()->clear_warning_info(thd->query_id);
      if (!(res= Events::drop_event(thd,
                                    &lex->spname->m_db, &lex->spname->m_name,
                                    lex->if_exists())))
        my_ok(thd);
    }
    break;
  }
```
If that’s not the appropriate place to handle that decision, please let me know where it should be.

File: sql/event_data_objects.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (vuvova) [reply sql/event_data_objects.cc:321]: as you like. Just remember that any unrelated refactoring should be done in a separate commit. Same PR is fine :) but a separate commit.

File: sql/sql_yacc.yy (hunk 12 above)
[the remark this answers is not in the document]
Comment (vuvova) [reply sql/sql_yacc.yy:13262]: > What do you suggest? Maybe make this decision in `mysql_create_or_drop_trigger()`?

I think this is a good idea. `add_table_for_trigger_internal` starts from
```c++
  if (check_trn_exists(&trn_path)) {
    if (if_exists) {
      push_warning_printf(thd, Sql_condition::WARN_LEVEL_NOTE,
                          ER_TRG_DOES_NOT_EXIST,
                          ER_THD(thd, ER_TRG_DOES_NOT_EXIST));
      *table= NULL;
      DBUG_RETURN(FALSE);
    }
    my_error(ER_TRG_DOES_NOT_EXIST, MYF(0));
    DBUG_RETURN(TRUE);
  }
```
you can, for example, move error reporting to the caller and `mysql_create_or_drop_trigger` can call drop_event.

Please, add tests for CREATE OR REPLACE, CREATE IF NOT EXISTS, and DROP IF EXISTS. Particularly, when a trigger of one type exists, but you create (with OR REPLACE or IF NOT EXISTS) a trigger of a different type

File: mysql-test/main/trigger.test [hunk 14]
```diff
@@ -2237,15 +2237,15 @@ create trigger t1_bi before insert on t1 for each row begin end;
...
+--error 1360,1192
 drop trigger t1_bi;
 
 lock tables t1 read;
 --error ER_TABLE_NOT_LOCKED_FOR_WRITE
 create trigger t1_bi before insert on t1 for each row begin end;
 --error ER_TABLE_NOT_LOCKED_FOR_WRITE
 create trigger t1_bi before insert on t1 for each row begin end;
---error ER_TRG_DOES_NOT_EXIST
+--error ER_LOCK_OR_ACTIVE_TRANSACTION
```
[the remark this answers is not in the document]
Comment (andremralves) [reply mysql-test/main/trigger.test:2248]: Previously, only DROP TRIGGER was being executed, and it worked fine with locked tables.

Now, when I run DROP TRIGGER and the trigger isn't found, DROP EVENT is executed instead.

However, there's a locked table, which causes the error ER_LOCK_OR_ACTIVE_TRANSACTION during DROP EVENT.

One alternative could be to don't even call DROP EVENT if there are locked tables, preventing this error. Is that possible?

File: mysql-test/main/trigger.test [hunk 15]
```diff
@@ -2237,15 +2237,15 @@ create trigger t1_bi before insert on t1 for each row begin end;
 create trigger t1_bi2 before insert on t1 for each row begin end;
 drop trigger t1_bi;
 drop trigger t1_bi2;
---error ER_TRG_DOES_NOT_EXIST
+--error 1360,1192
```
[the remark this answers is not in the document]
Comment (andremralves) [reply mysql-test/main/trigger.test]: Drop event was failing because there was a locked table. But it was inconsistent and I was getting both errors ER_TRG_DOES_NOT_EXIST or ER_LOCK_OR_ACTIVE_TRANSACTION. I fixed it by only running drop_event if there isn't an active locked table.

File: mysql-test/main/trigger.test (hunk 14 above)
[the remark this answers is not in the document]
Comment (andremralves) [reply mysql-test/main/trigger.test]: I did it, and it worked. Now, I'm not calling drop_event if there is an active locked table.

File: sql/event_data_objects.cc [hunk 16]
```diff
@@ -595,6 +599,10 @@ Event_queue_element::load_from_row(THD *thd, TABLE *table)
   on_completion= (ptr[0]=='D'? Event_parse_data::ON_COMPLETION_DROP:
                                Event_parse_data::ON_COMPLETION_PRESERVE);
 
+  if(!table->field[ET_FIELD_EVENT_KIND]->is_null()) {
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/event_data_objects.cc]: Removed

File: sql/event_parse_data.h [hunk 17]
```diff
@@ -54,6 +54,14 @@ class Event_parse_data : public Sql_alloc
     ON_COMPLETION_PRESERVE
   };
 
+  enum enum_event_kind
+  {
+    SCHEDULE = 1,
```
[the remark this answers is not in the document]
Comment (andremralves) [reply sql/event_parse_data.h:59]: I did it this way, considering that table enums are one-indexed. Zero would represent an empty string, I guess, but that will probably never happen.

[the remark this answers is not in the document]
Comment (andremralves) [reply sql/event_parse_data.h:59]: I based the implementation on this image:

image

So event_kind would be:

| Index  | Value  |
|---|---|
| NULL | NULL |
| 0 | '' |
| 1 | SCHEDULE |
| 2 | STARTUP |
| 3 | SHUTDOWN |
